samples/example_vptr/example_vptr.cpp

```diff
@@ -0,0 +1,115 @@
+/***************************************************************************
+ *
+ *  Copyright (C) 2017 Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  Codeplay's ComputeCpp SDK
+ *
+ *  example_vptr.cpp
+ *
+ *  Description:
+ *    Sample code that demonstrates the use of the virtual pointer intrface in
+ *SYCL on matrix addition.
+ *
+ **************************************************************************/
+
+#include <CL/sycl.hpp>
+
+#include <iostream>
+
+#include "virtual_ptr.hpp"
+
+using namespace cl::sycl;
+
+const size_t N = 100;
+const size_t M = 150;
+
+/* This sample allocates three device-only matrices, using the virtual pointer
+ * and SYCLmalloc. It initalises the first two in parallel on the device. After
+ * that, it adds them together, storing the result in the third matrix. It then
+ * verifies the result on the host by using:
+ *  - pointer arithmetic on the virtual pointer to index the matrix
+ *  - host accessor to gain access to the data. */
+int main() {
+  {
+    queue myQueue;
+    cl::sycl::codeplay::PointerMapper pMap;
+
+    /* Allocate the matrices using SYCLmalloc. a, b and c are virtual pointers,
+     * pointing to device buffers.
+     */
+    float* a = static_cast<float*>(SYCLmalloc(N * M * sizeof(float), pMap));
+    float* b = static_cast<float*>(SYCLmalloc(N * M * sizeof(float), pMap));
+    float* c = static_cast<float*>(SYCLmalloc(N * M * sizeof(float), pMap));
+
+    /* This kernel will initialise the buffer pointed to by a. The accessor "A"
+     * has write access. We retrieve it directly from the PointerMapper, using
+     * the virtual pointer.*/
+    myQueue.submit([&](handler& cgh) {
+      auto A = pMap.get_access<access::mode::discard_write,
+                               access::target::global_buffer, float>(a, cgh);
```

Comment:
I guess `, float` can be inferred here and is redundant in this example.
Could it be used then to do a `reinterpret_cast` when required?
Then you can remove all the `float` remaining in the example.

Reply:
Unfortunately, as the code is at the moment, we need to provide the data type, as we don not save it when using SYCLmalloc and it cannot be inferred. It would be great if we could infer it though. I will look into doing this in another MR.